Support Crate DB Authentication

modularTaco opened the pull request:
## Proposed changes

Related to issue #474 

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [x] I have added tests that prove my fix is effective or that my
      feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

Added a second crate docker container with authentication enabled and added this to all tests which are running on timescale an crate. This guarantees there are no missing permissions for the QL user in CrateDB.  
Note: Some tests failed randomly on my machine. Tried it on a clean Ubuntu VM and had the same problem. Running the tests again fixes it somestimes, but sometimes not. This was also the case before my changes. The error is always a `SchemaUnknownException` like `crate.client.exceptions.ProgrammingError: SchemaUnknownException[Schema 'mttest' unknown]` e.g. in test `test_cache_insert_with_tenant[crate]`. This happens with the `crate` as well as the `crate-auth` container.  
Hopefully the tests run fine in CircleCI.



[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

modularTaco commented:
I have read the CLA Document and I hereby sign the CLA

chicco785 commented:
@modularTaco i think it's better to rebase and merge this pr after @modularTaco  reworks of the test.
fyi the issue we are working on is essentially related to cratedb latency in consolidating data. depening on the hw capacities, test fails because we query the data before crate consolidates it (cf #574)

modularTaco commented:
> @modularTaco i think it's better to rebase and merge this pr after @modularTaco reworks of the test. fyi the issue we are working on is essentially related to cratedb latency in consolidating data. depening on the hw capacities, test fails because we query the data before crate consolidates it (cf #574)

Hey @chicco785 ,

thank you for the fast response!

You mean after @c0c0n3 reworks of the test?  
Makes totally sense to me. 

Do you have any more details on this issue (probably cratedb documention)? We experience this as well in the logs of our QL deployment from time to time.

chicco785 commented:
> Hey @chicco785 ,
> 
> thank you for the fast response!
> 
> You mean after @c0c0n3 reworks of the test? Makes totally sense to me.

exactly

> 
> Do you have any more details on this issue (probably cratedb documentation)? We experience this as well in the logs of our QL deployment from time to time.

maybe @c0c0n3 can provide some references, but essentially this is related to the shared nothing and eventual consistent model adopted by crate.

chicco785 commented:
@modularTaco can you sign the CLA?
I will now go through the code and give you my comments (if any)

chicco785 commented on setup_dev_env.sh:
```diff
@@ -28,6 +28,7 @@ export MONGO_HOST=$LH
 
 export QL_HOST=$LH
 export CRATE_HOST=$LH
+export CRATE_AUTH_HOST=$LH
```
i had a quick look at the whole changes:
i would not duplicate the configuration auth / no auth, this makes test longer for no reason.
so let's stick just add auth to the already deployed crate.

wdyt?

modularTaco replied:
I thought running the tests twice, with "normal" crate and with crate with authentication enabled, would be better to ensure the changes have no impact in environments with unauthenticated crates and to make it kind of easier to debug (if the test fails in crate and crate-auth, probably missing permissions is not the reason for this. if it works in crate but fails in crate-auth, it may be related to permissions). And additionally to make it easier to run the tests with older crate version (pre 4.5, where authentication was an enterprise feature), in that case the cratedb version env just needs to be set and the crate-auth tests deselected.

But i'm not sure if this is enough benefit to justify the longer test run

chicco785 commented on src/tests/docker-compose.yml:
```diff
@@ -59,11 +60,24 @@ services:
       - PG_HOST=timescale
       - PG_PASS=*
 
+  crate-auth:
```
se comment above

chicco785 commented on src/translators/crate.py:
```diff
@@ -42,13 +42,15 @@
     NGSI_TO_SQL = NGSI_TO_SQL
 
-    def __init__(self, host, port=4200, db_name="ngsi-tsdb"):
+    def __init__(self, host, port=4200, db_name="ngsi-tsdb", username=None, password=None):
```
i don't recall where, but i experienced the case that i had to put 'crate' as username if using the driver with username to connect. if that the case, maybe better 'crate' as default value.

modularTaco replied:
as far as i know (just looking in my memory, not in the docs) crate uses "crate" as username per default if the connection is done via the crate driver. if the connection uses the postgres driver "crate" needs to be used as username explicitly.

As the tests run fine with the (non-auth) crate container i think it works well.

i can set the default to crate instead of None and run the tests again, but i do not think that it'll make any difference

chicco785 reviewed (changes requested):
see my comments, and of course you need as well to rebase (should be straight forward)

modularTaco commented:
> @modularTaco can you sign the CLA? I will now go through the code and give you my comments (if any)

I think i have?

![image](https://user-images.githubusercontent.com/37046961/145093526-0cb7ddad-05f9-47ad-a8d1-493cff3f6451.png)

modularTaco commented:
> see my comments, and of course you need as well to rebase (should be straight forward)

Thank you for your quick response! Will rebase the branch tomorrow.

chicco785 commented:
@modularTaco i forced the rebase to see if error get solved, unfortunately not.
again i don't think you need two different set of tests. you can simply add a specific test, in translator, by enabling and disabling authentication.

modularTaco commented:
> @modularTaco i forced the rebase to see if error get solved, unfortunately not. again i don't think you need two different set of tests. you can simply add a specific test, in translator, by enabling and disabling authentication.

Yes, unfortunately the error is still Relation unknown and happens in both targets, crate and crate-auth (so i guess it's still related to the tests and not to my changes).

What do you mean by "you can simply add a specific test, in translator, by enabling and disabling authentication."? To enable authentication i have to start crate with different params, so i need to shut crate down, change the startup params (or start another prepared container in the docker-compose file), wait until crate with auth has started, configure a dedicated user, password and permissions, run some kind of test (what do you suggest here? Just inserting some data?), shutdown crate with auth and start the normal crate again. If this is what you meant i can try to implement it later this week (and remove the duplicate tests for crate-auth)

chicco785 commented:
> > @modularTaco i forced the rebase to see if error get solved, unfortunately not. again i don't think you need two different set of tests. you can simply add a specific test, in translator, by enabling and disabling authentication.
> 
> Yes, unfortunately the error is still Relation unknown and happens in both targets, crate and crate-auth (so i guess it's still related to the tests and not to my changes).
> 
> What do you mean by "you can simply add a specific test, in translator, by enabling and disabling authentication."? To enable authentication i have to start crate with different params, so i need to shut crate down, change the startup params (or start another prepared container in the docker-compose file), wait until crate with auth has started, configure a dedicated user, password and permissions, run some kind of test (what do you suggest here? Just inserting some data?), shutdown crate with auth and start the normal crate again. If this is what you meant i can try to implement it later this week (and remove the duplicate tests for crate-auth)

I think we can look into two options:
1. always use authenticated;
2. just test that auth works once in translator tests; in this case, hopefully, you are able to pass additional config via an env variable to the same docker service to enable auth

modularTaco commented:
hey @chicco785 ,

i removed all the duplicated tests and the second docker container.  
Instead i reconfigured the crate container to enable authentication and allow trust auth for everyone (so all tests can connect with user crate and no password) as follows:

```
    command: crate -Cauth.host_based.enabled=true
      -Cauth.host_based.config.0.user=crate
      -Cauth.host_based.config.0.address=_local_
      -Cauth.host_based.config.0.method=trust
      -Cauth.host_based.config.1.user=crate
      -Cauth.host_based.config.1.address=0.0.0.0/0
      -Cauth.host_based.config.1.method=trust
      -Cauth.host_based.config.99.method=password -Ccluster.name=democluster
      -Chttp.cors.enabled=true -Chttp.cors.allow-origin="*"
```

Afterwards i create a dedicated quantumleap user in `translators/tests/conftest.py` (as it is now, just in the `crate` instead of the `crate-auth` container).

This works fine, all translator tests pass.

In addition i've added a test to ensure the crate authentication works as expected:

```
@pytest.mark.parametrize("translator", [pytest.lazy_fixture('crate_auth_translator')], ids=["crate-auth"])
def test_auth_is_not_superuser(translator):
    stmt = "select current_user"
    translator.cursor.execute(stmt)
    res = translator.cursor.fetchall()
    current_user = res[0][0]
    assert current_user == 'quantumleap'
```

Running this with `pytest -k auth` (so this is the only test which will be run) works fine.

But if i run all tests, this is the only test which fails because of 

```
E       AssertionError: assert 'crate' == 'quantumleap'
E         - crate
E         + quantumleap
```

And at the moment i have no idea why it isn't using the correct pytest fixture in this case... Do you have any ideas here? 

I can push the changes to this branch if you wish, haven't done it yet because i didn't want to push commits with failing tests

chicco785 commented:
> hey @chicco785 ,
> 
> i removed all the duplicated tests and the second docker container. Instead i reconfigured the crate container to enable authentication and allow trust auth for everyone (so all tests can connect with user crate and no password) as follows:
> 
> ```
>     command: crate -Cauth.host_based.enabled=true
>       -Cauth.host_based.config.0.user=crate
>       -Cauth.host_based.config.0.address=_local_
>       -Cauth.host_based.config.0.method=trust
>       -Cauth.host_based.config.1.user=crate
>       -Cauth.host_based.config.1.address=0.0.0.0/0
>       -Cauth.host_based.config.1.method=trust
>       -Cauth.host_based.config.99.method=password -Ccluster.name=democluster
>       -Chttp.cors.enabled=true -Chttp.cors.allow-origin="*"
> ```
> 
> Afterwards i create a dedicated quantumleap user in `translators/tests/conftest.py` (as it is now, just in the `crate` instead of the `crate-auth` container).
> 
> This works fine, all translator tests pass.
> 
> In addition i've added a test to ensure the crate authentication works as expected:
> 
> ```
> @pytest.mark.parametrize("translator", [pytest.lazy_fixture('crate_auth_translator')], ids=["crate-auth"])
> def test_auth_is_not_superuser(translator):
>     stmt = "select current_user"
>     translator.cursor.execute(stmt)
>     res = translator.cursor.fetchall()
>     current_user = res[0][0]
>     assert current_user == 'quantumleap'
> ```
> 
> Running this with `pytest -k auth` (so this is the only test which will be run) works fine.
> 
> But if i run all tests, this is the only test which fails because of
> 
> ```
> E       AssertionError: assert 'crate' == 'quantumleap'
> E         - crate
> E         + quantumleap
> ```
> 
> And at the moment i have no idea why it isn't using the correct pytest fixture in this case... Do you have any ideas here?
> 
> I can push the changes to this branch if you wish, haven't done it yet because i didn't want to push commits with failing tests

yes please update the branch, I will then review code :) if you add me to you project, I can also modify code as needed.

modularTaco commented:
> yes please update the branch, I will then review code :) if you add me to you project, I can also modify code as needed.

ok perfect, thank you

I've added you to the project and updated the branch

chicco785 commented:
> But if i run all tests, this is the only test which fails because of
> 
> ```
> E       AssertionError: assert 'crate' == 'quantumleap'
> E         - crate
> E         + quantumleap
> ```

to speed up sql tasks, connection is cached (see https://github.com/modularTaco/ngsi-timeseries-api/blob/dev-crate-auth/src/translators/crate.py#L93), i solved by disposing the connection in the fixture (see https://github.com/modularTaco/ngsi-timeseries-api/blob/dev-crate-auth/src/conftest.py#L195). this may slow down tests, let's see, if that the case i will dispose the connection only in the auth tests.

now, i also refactored few things, and because of that integration tests brakes :) I will look into that tomorrow.

chicco785 commented on src/tests/docker-compose.yml:
```diff
@@ -59,11 +59,17 @@ services:
     image: crate:${CRATE_VERSION:-4.5.1}
-    command: crate -Cauth.host_based.enabled=false
-      -Ccluster.name=democluster -Chttp.cors.enabled=true -Chttp.cors.allow-origin="*"
+    command: crate -Cauth.host_based.enabled=true
```
this is the reason why bc test are failing, ql 0.7.5 of course does not support auth

chicco785 commented:
@modularTaco that's should be it :) time of test didn't increase so we can leave disposal where it is
have a look, if you don't have other things to change / add, we can merge it

chicco785 reviewed (approved):
LGTM

modularTaco commented:
Wow okay, a lot more than what i've expected

-
    > to speed up sql tasks, connection is cached 

    As i used different instances of the crate translator class, i thought the caching was related to each instance and therefore should not interfere. But i think i've missed the shared state of the connection manager here

- personally i'd prefer `CRATE_USERNAME` instead of `CRATE_DB_USERNAME`, as it refers to multiple databases/schemas in crate and not only to one, just like `CRATE_HOST` and `CRATE_PORT` do either. In case of timescale it looks like a parameter `POSTGRES_DB_NAME` is given as well, so the credentials will most likely be only valid for this specific database. Never used QL with timescale, so i'm not sure.

- just removed the TODO comment in the test, as you've already solved it


Thank you for the nice and extensive support! Looks very good to me now with your fixes

chicco785 commented:
> Wow okay, a lot more than what i've expected
> 
> * > to speed up sql tasks, connection is cached
>   
>   
>   As i used different instances of the crate translator class, i thought the caching was related to each instance and therefore should not interfere. But i think i've missed the shared state of the connection manager here

given that at the time being we don't support different instances of crate or timescale, the connection manager is shared for db instance. this, of course, may change in future.

> * personally i'd prefer `CRATE_USERNAME` instead of `CRATE_DB_USERNAME`, as it refers to multiple databases/schemas in crate and not only to one, just like `CRATE_HOST` and `CRATE_PORT` do either. In case of timescale it looks like a parameter `POSTGRES_DB_NAME` is given as well, so the credentials will most likely be only valid for this specific database. Never used QL with timescale, so i'm not sure.

i think it works the same, the purpose of `DB` was to allow - if needed by some specific operations - to also have a superuser password.

> * just removed the TODO comment in the test, as you've already solved it
> 
> Thank you for the nice and extensive support! Looks very good to me now with your fixes

my pleasure, we welcome contributions, and do our best to guide contributors through them, of course depending on zillions of factors... but hey, that's life :)

Cerfoglg commented on src/conftest.py:
```diff
@@ -128,23 +130,27 @@ def clean_crate():
 
         def insert(self, entities,
-                   fiware_service=None, fiware_servicepath=None):
+                    fiware_service=None, fiware_servicepath=None):
```
Extra spaces here?

chicco785 replied:
could be... next autopep8 will fix, unfortunately, on external pull requests we can't enforce autopep8 since we don't own the repo...

Cerfoglg reviewed (approved):
Looks good to me
